rialto_airflow/harvest/pubmed.py

```diff
@@ -0,0 +1,196 @@
+    count = 0
+    stop = False
+
+    with jsonl_file.open("w") as jsonl_output:
+        with get_session(snapshot.database_name).begin() as select_session:
+            # get all authors that have an ORCID
+            for author in (
+                select_session.query(Author).where(Author.orcid.is_not(None)).all()  # type: ignore
+            ):
+                if stop is True:
+                    logging.info(f"Reached limit of {limit} publications stopping")
+                    break
+
+                pmids = pmids_from_orcid(author.orcid)
+                if not pmids:
+                    logging.info(f"No publications found for {author.orcid}")
+                    continue
+
+                for pubmed_pub in publications_from_pmids(pmids):
+                    count += 1
+                    if limit is not None and count > limit:
+                        stop = True
+                        break
+
+                    doi = normalize_doi(get_doi(pubmed_pub))
+
+                    with get_session(snapshot.database_name).begin() as insert_session:
+                        # if there's a DOI constraint violation we need to update instead of insert
+                        pub_id = insert_session.execute(
+                            insert(Publication)
+                            .values(
+                                doi=doi,
+                                pubmed_json=pubmed_pub,
+                            )
+                            .on_conflict_do_update(
+                                constraint="publication_doi_key",
+                                set_=dict(pubmed_json=pubmed_pub),
+                            )
+                            .returning(Publication.id)
+                        ).scalar_one()
+
+                        # a constraint violation is ok here, since it means we
+                        # already know that the publication is by the author
+                        insert_session.execute(
+                            insert(pub_author_association)
+                            .values(publication_id=pub_id, author_id=author.id)
+                            .on_conflict_do_nothing()
+                        )
+
+                        jsonl_output.write(json.dumps(pubmed_pub) + "\n")
+
+    return jsonl_file
+
+
+def pmids_from_orcid(orcid):
+    """
+    Returns PMIDs associated with a given ORCID.
+    """
+    # Pubmed doesn't want the full ORCID URIs which are stored in User table
+    if m := re.match(r"^https?://orcid.org/(.+)$", orcid):
+        orcid = m.group(1)
+
+    return _pubmed_search_api(f"{orcid}[auid]")
+
+
+def publications_from_pmids(pmids):
+    """
+    Returns full pubmed records given a list of PMIDs.
+    """
+    if not pmids or not isinstance(pmids, list):
+        return None
+
+    query = "id=" + "&id=".join(pmids)
+
+    http = requests.Session()
+
+    full_url = (
+        f"{BASE_URL}{FETCH_PATH}&api_key={os.environ.get('AIRFLOW_VAR_PUBMED_KEY')}"
```

Comment:
Maybe the key could be looked up at the top of the file where you are setting some constants instead of getting from the environment over and over?

Reply:
created a method to fetch the key